src/material/datepicker/calendar-body.ts

```diff
@@ -144,4 +175,75 @@ export class MatCalendarBody implements OnChanges {
       });
     });
   }
+
+  /** Gets whether the calendar is currently selecting a range. */
+  _isRange(): boolean {
+    return this.startValue !== this.endValue;
+  }
+
+  /** Gets whether a value is within the currently-selected range. */
+  _isInRange(value: number): boolean {
+    if (!this._isRange() || value < this.startValue) {
+      return false;
+    }
+
+    return value <= this.endValue || value <= this._hoveredValue;
+  }
+
+  /** Event handler for when the user's pointer enters an element inside the calendar body. */
+  private _mouseenter = (event: MouseEvent) => {
+    // We only need to hit the zone when we're selecting a range, we
+    // have a start value without an end value and we've hovered over a date cell.
+    if (!event.target || !this._isRange() || !this.startValue || this.endValue) {
+      return;
+    }
+
+    const cell = this._getCellFromElement(event.target as HTMLElement);
+
+    if (cell) {
+      this._ngZone.run(() => {
+        this._hoveredValue = cell.value;
+        this._changeDetectorRef.markForCheck();
+      });
+    }
+  }
+
+  /** Event handler for when the user's pointer leaves an element inside the calendar body. */
+  private _mouseleave = (event: MouseEvent) => {
+    // We only need to hit the zone when we're selecting a range.
+    if (this._isRange() && this._hoveredValue !== -1) {
+      // Only reset the hovered value when leaving cells. This looks better, because
+      // we have a gap between the cells and the rows and we don't want to remove the
+      // range just for it to show up again when the user moves a few pixels to the side.
+      if (event.target && (event.target as HTMLElement).nodeName === 'TD') {
```

Comment:
I see a lot of `=== 'TD'` can we put this in a helper method to make the code more readable?